brewtils/request_handling.py

```diff
@@ -0,0 +1,397 @@
+    parse_exception_as_json,
+    RepublishRequestException,
+    RestConnectionError,
+    RestClientError,
+    RequestProcessingError,
+)
+from brewtils.models import Request
+from brewtils.schema_parser import SchemaParser
+
+
+class RequestProcessor(object):
+    def __init__(
+        self,
+        target,
+        updater,
+        validation_funcs=None,
+        logger=None,
+        unique_name=None,
+        max_workers=None,
+    ):
+        self.logger = logger or logging.getLogger(__name__)
+
+        self._target = target
+        self._updater = updater
+        self._unique_name = unique_name
+        self._validation_funcs = validation_funcs or []
+        self._pool = ThreadPoolExecutor(max_workers=max_workers)
+
+    def on_message_received(self, message, headers):
+        """Callback function that will be invoked for received messages
+
+        This will attempt to parse the message and then run the parsed Request through
+        all validation functions that this RequestProcessor knows about.
+
+        If the request parses cleanly and passes validation it will be submitted to this
+        RequestProcessor's ThreadPoolExecutor for processing.
+
+        Args:
+            message: The message string
+            headers: The header dictionary
+
+        Returns:
+            A future that will complete when processing finishes
+
+        Raises:
+            DiscardMessageException: The request failed to parse correctly
+            RequestProcessException: Validation failures should raise a subclass of this
+        """
+        request = self._parse(message)
+
+        for func in self._validation_funcs:
+            func(request)
+
+        # This message has already been processed, all it needs to do is update
+        if request.status in Request.COMPLETED_STATUSES:
+            return self._pool.submit(self._updater.update_request, request, headers)
+        else:
+            return self._pool.submit(
+                self.process_message, self._target, request, headers
+            )
+
+    def process_message(self, target, request, headers):
+        """Process a message. Intended to be run on an Executor.
+
+        :param target: The object to invoke received commands on.
+            (self or self.client)
+        :param request: The parsed Request
+        :param headers: Dictionary of headers from the
+            `brewtils.request_consumer.RequestConsumer`
+        :return: None
+
+        """
+        request.status = "IN_PROGRESS"
+        self._updater.update_request(request, headers)
+
+        try:
+            # Set request context so this request will be the parent of any
+            # generated requests and update status We also need the host/port of
+            #  the current plugin. We currently don't support parent/child
+            # requests across different servers.
+            import brewtils.plugin
+
+            brewtils.plugin.request_context.current_request = request
```

Comment:
This seems odd. Can we not import this earlier? It seems weird that we are importing every time any request is executed. Also, if we want that request context stuff to be hidden, we should add a method call or something maybe?

Reply:
Not sure what happened with the import, I moved it to the top. I'm not sure what you mean about making the request context stuff hidden?